AdaptiveRoads/UI/QuayRoads/QuayRoadsPanelField.cs

```diff
@@ -0,0 +1,111 @@
+using AdaptiveRoads.Manager;
+using ColossalFramework.UI;
+using KianCommons;
+using KianCommons.UI;
+using static KianCommons.EnumBitMaskExtensions;
+using static KianCommons.EnumerationExtensions;
+using static KianCommons.ReflectionHelpers;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using AdaptiveRoads.Data.QuayRoads;
+using UnityEngine;
+using static UnityEngine.Object;
+using System.Globalization;
+using ColossalFramework;
+using AdaptiveRoads.Util;
+
+namespace AdaptiveRoads.UI.QuayRoads {
+    class QuayRoadsPanelField {
+        private NetInfo netInfo_;
+        private int sectionIndex_;
+        private FieldInfo fieldInfo_;
+        private IConvertible? flag_;
+        private RoadEditorPanel parentPanel_;
+
+        public QuayRoadsPanelField(NetInfo netInfo, int sectionIndex, FieldInfo fieldInfo, IConvertible flag, UIPanel parent, RoadEditorPanel parentPanel) {
+            netInfo_ = netInfo;
+            sectionIndex_ = sectionIndex;
+            fieldInfo_ = fieldInfo;
+            flag_ = flag;
+            parentPanel_ = parentPanel;
+            if (flag_ is not null) {
+
+                Log.Debug(flag.ToString());
+                Log.Debug(flag.ToUInt64().ToString());
+                var propertyCheckbox = parent.AddUIComponent<UICheckBoxExt>();
+                propertyCheckbox.Label = "";
+                propertyCheckbox.width = 20f;
+                propertyCheckbox.isChecked = (assetValue_ as IConvertible).IsFlagSet(flag);
+                propertyCheckbox.eventCheckChanged += (_, _isChecked) => {
+                    //TODO: find a better / more robust way to do this.
+                    if (_isChecked) {
+                        if (Enum.GetUnderlyingType(fieldInfo_.FieldType) == typeof(Int32)) {
+                            assetValue_ = (assetValue_ as IConvertible).ToInt32(CultureInfo.InvariantCulture) | flag.ToInt32(CultureInfo.InvariantCulture);
+                        } else {
+                            assetValue_ = (assetValue_ as IConvertible).ToInt64() | flag.ToInt64();
+                        }
+                    } else {
+                        if (Enum.GetUnderlyingType(fieldInfo_.FieldType) == typeof(Int32)) {
+                            assetValue_ = (assetValue_ as IConvertible).ToInt32(CultureInfo.InvariantCulture) & ~flag.ToInt32(CultureInfo.InvariantCulture);
+                        } else {
+                            assetValue_ = (assetValue_ as IConvertible).ToInt64() &~flag.ToInt64();
+                        }
+                    }
+                };
+            } else {
+                // TODO: right now, this assumes everything that is not a flags enum is a float
+                // TODO: find a better way to create a text field - maybe something similar to UICheckBoxExt from KianCommons?
```

Comment:
I tried duplicating UICheckBoxExt to create a UITextFieldExt, but couldn't figure out how to make it interactive.

Reply:
look at MiniPanelTextField in this project. there are some numeric textfield and float text field stuff there too. maybe we can move part of the code to kian commons.
 
BTW you might be able to improve this if you look at MiniPanelFloatField MiniPanelNumberField MiniPanelTextField